### DISCUSSION TITLE
Chakra UI & React Router: active style

### DISCUSSION THREAD

**Discussion by tutods** [chakra-ui/chakra-ui#4777] [Q&A]
Hi Guys
I have a menu item with text and icon, I using the React Router DOM to manage the Routes, and I need customize the menu item when the route as active, but I only got the customization on container (in case `<Link as={NavLink}>`), I can't customize the style on Icon.

Anyone can help me?
Is possible do that?

**Comment by adenta** [comment]
What did you end up doing?

**Comment by adenta** [comment]
Lol this "works"

```
    <Link
      as={(props) => (
        <NavLink
          {...props}
          style={({ isActive }) => {
            return { fontWeight: isActive ? 700 : 400 };
          }}
        />
      )}
      
      to={to}
    >
      {children}
    </Link>
```

Pretty sure the problem with this is the inline styles, which I believe are an esoteric security concern.

**Comment by keithburgie** [comment]
You can target children with the Sass-style syntax below
```javascript
<RouterNavLink
  to={page.slug}
  _activeLink={{
    fontWeight: "bold",
    "& > svg": {
      background: "red",
    },
  }}
>
  {page.title}
  <StarIcon />
</RouterNavLink>
```
`&` means "this component
`>` is a direct child selector. It would be optional in this case.
`svg` is the element type of the icon.
So, `& > svg` means "a direct child of this component that is an svg`

You can figure out what you need to target by inspecting within your browser:
![Screen Shot 2021-12-19 at 6 38 57 AM](https://user-images.githubusercontent.com/96078669/146673612-277155cd-33a9-41cc-86d5-d212e36329f0.png)

So in this case, if you wanted, you could target the svg (like above) or the icon className: `& > .chakra-icon`

I've added some alignment styles here but this would be the effect:
![Screen Shot 2021-12-19 at 6 34 02 AM](https://user-images.githubusercontent.com/96078669/146673405-ad057372-a635-45ac-ba2a-30c77b9d2dd7.png)

**Reply by madrimovDev** [reply]
thank you bro 👍

**Comment by saronlujan** [comment]
Work for me...

```
<Link as={NavLink} end to="/my-path" _activeLink={{fontWeight:"bold"}}>
         My Link
 </Link>
```
or 

```
<Link as={NavLink} end to="my-path"  _hover={{textDecoration: 'none'}} _focus={{outline:'none'}} _activeLink={{color:'#666'}} >
     {({ isActive }) => ((
           isActive ? 
                <Flex justifyContent="space-between">
                     <Text>{children}</Text> <Icon as={FiChevronRight} />
                </Flex> 
            : 
                 <Text>my link</Text>
            ))}
</Link>
```

If anyone has a simpler way for the second option...

**Reply by keithburgie** [reply]
I think this is the simpler option #2 that you’re looking for. There’s no harm in having your inactive link wrapped in another box. 
```
<Link as={NavLink} end to="my-path" _hover={{textDecoration: 'none'}} _focus={{outline:'none'}} _activeLink={{color:'#666'}} >
  {({ isActive }) => (
    <Flex justifyContent="space-between">
      <Text>{children}</Text>
      {isActive && <Icon as={FiChevronRight} marginLeft={2} />}
    </Flex> 
  )}
</Link> 
```

**Comment by Mayvis** [comment]
@saronlujan thank you, I follow your suggestion, it's work, but still have some TypeScript issue.
Does anyone know how to fix this or maybe just create a button like router link? 

> Type '({ isActive }: { isActive: any; }) => Element' is not assignable to type 'ReactNode'.ts(2322)

![Screen Shot 2022-06-24 at 11 33 39 AM](https://user-images.githubusercontent.com/12667910/175457748-31740244-1302-4f61-929a-5500699c6e85.png)

**Reply by keithburgie** [reply]
@Mayvis, comment out your `leftIcon` and see if the TypeScript error goes away.

**Reply by saronlujan** [reply]
Hi! Unn... I didn't get to use it with TS. But if you find the problem, send it to us.

**Reply by Mayvis** [reply]
@saronlujan @keithburgie sorry for late replay, this issue still exists.
I just use `npx create-react-app react-app --template typescript` to create empty react typescript project, then install chakra-ui and react-router-dom. 

"dependencies": {
    "@chakra-ui/react": "^2.2.4",
    "@emotion/react": "^11.9.3",
    "@emotion/styled": "^11.9.3",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.5.2",
    "@types/node": "^16.11.45",
    "@types/react": "^18.0.15",
    "@types/react-dom": "^18.0.6",
    "framer-motion": "^6.5.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.3.0",
    "react-scripts": "5.0.1",
    "typescript": "^4.7.4",
    "web-vitals": "^2.1.4"
  }

![Screen Shot 2022-07-23 at 11 19 48 AM](https://user-images.githubusercontent.com/12667910/180588908-05a0df6a-77f9-4fbe-a219-b80a68816e27.png)

**Reply by tyteen4a03** [reply]
Can confirm this is still an issue.

**Reply by fransfilastap** [reply]
<img width="793" alt="image" src="https://user-images.githubusercontent.com/10008396/196893110-cfa84672-f210-4e22-b38b-159468cfe50d.png">

I have the same issue. Typescript failed to recognize props : {isActive:boolean,isPending:boolean}. In development, it will be fine but will be failed to in compile process. 

Need a workaround for this issue.

**Reply by fransfilastap** [reply]
I ended up using [saronlujan](https://github.com/saronlujan) first solution. 

```
<Link
    as={NavLink}
    _activeLink={{
            color:useColorModeValue('violet.50','white'),
            bgColor:useColorModeValue('violet.5','gray.700')
            }}>
    Text
</Link>
```

**Reply by sohamtamboli** [reply]
@Mayvis Hi, have you found a solution to this typescript error?

**Reply by Mayvis** [reply]
@sohamtamboli Nope, I hadn't found a solution. This is Chakra ui issue at that time😅. 
Instead using chakra ui Link component, you can build your own custom component to handle this situation.

**Comment by laduke** [comment]
this seems to work too

    <Box as={NavLink} to={'my-path'}>
        {({ isActive }) => (
            <Button isActive={isActive} ...

**Comment by wc2184** [comment]
Building off saronlujan's answer, one has a red color when it's the current link active, none if not
```
<NavLink active to="/home">
                {        
    ({ isActive }) => isActive ?  <Button colorScheme="red">Home</Button> : <Button>Home</Button>   
                }
</NavLink>

**Comment by LucasRodriguesOliveira** [comment]
It seems that it has been a long time since this issue was posted, probably in older Chakra versions (I'm using v3.26.0), but I'd like to contribute with my approach. In the documentation of both react-router and Chakra I've found that in the `NavLink` component from `react-router` automatically adds a property `aria-current="page"`, also, `Link` component from `Chakra` has a property `_currentPage` where you can add any style, adding `asChild` to the `Link` component, we can achieve the expected result with the following code:

```tsx
import type { FC } from 'react';
import { Link as ChakraLink } from '@chakra-ui/react';
import { NavLink as RouterNavLink } from 'react-router';

export type NavLinkProps = {
  label: string;
  href: string;
};

export const NavLink: FC<NavLinkProps> = ({ label, href }) => {
  return (
    <ChakraLink _currentPage={{ color: 'red' }} asChild>
      <RouterNavLink to={href}>{label}</RouterNavLink>
    </ChakraLink>
  );
};
```
